website/blog/en/2021/clickhouse-v21.12-released.md

```diff
@@ -0,0 +1,60 @@
+date: '2021-12-13'
+author: '[Rich Raposa](https://github.com/rfraposa), [Alexey Milovidov](https://github.com/alexey-milovidov)'
+tags: ['company', 'community']
+---
+
+We're continuing our monthly release cadence and blog updates at[ ClickHouse, Inc](https://clickhouse.com/blog/en/2021/clickhouse-inc/). Let's highlight some of these new exciting new capabilities in 21.12:
+
+## Bool data type
+
+ClickHouse now natively supports a `bool` data type. Supported input values are “true”/”false”, “TRUE”/”FALSE” and “1”/”0”. By default, the text representation for CSV/TSV output is “true” and “false” but can be adjusted to anything else using the settings `bool_true_representation` and `bool_false_representation` (for example, “yes” and “no”).
+
+**How does this help you?**
+
+Native boolean data types exist today in other databases that are often integrated with ClickHouse, such as PostgreSQL. The `bool` data type in ClickHouse will make it more compatible with existing code and ease migration from other databases.
+
+
+## Partitions for File, URL, and HDFS storage
+
+When using the table engines file, url, and hdfs ClickHouse now supports partitions. When creating a table you can specify the partition key using the `PARTITION BY` clause e.g. `CREATE TABLE hits_files (&lt;columns>) ENGINE=File(TabSeparated) PARTITION BY toYYYYMM(EventDate)`.
+
+Similarly, when exporting data from ClickHouse using the file, url, and hdfs table functions you can now specify that the data is to be partitioned into multiple files using a `PARTITION BY` clause. For example, `INSERT INTO TABLE FUNCTION file(’&lt;path>/hits_{_partition_id}’, ‘TSV’, ‘&lt;table_format>’) PARTITION BY toYYYYMM(EventDate) VALUES &lt;values>’ will create as many files as there are unique month in the dataset.
+
+The S3 table function has supported partitioned writes since ClickHouse 21.10.
+
+**How does this help you?**
+
+This feature makes using tables backed by files on local file systems and remote file systems accessed over HTTP and HDFS more convenient and flexible. Especially for large datasets you might not want all data to be in one big file. This feature allows you to split one table into smaller files any way you want e.g. by month.
+
+
+## Table constraints
+
+When creating tables, you can now (optionally) specify constraints. A constraint tells ClickHouse that a column has a specific relationship to another column in the same table. For example, a string column might store the prefix of another column. Then, when a select query is trying to calculate the prefix on the original column, ClickHouse will rewrite the query to use the prefix column.
+
+For now, constraints are implemented as assumptions, that is, ClickHouse does not validate that they are correct and will not reject new data that violates them.
+
+This feature is disabled by default. To turn it on, enable `optimize_using_constraints`, `optimize_substitute_columns` and/or `optimize_append_index`.
+
```

Comment:
I think it's also worth to mention new secondary index of type `hypothesis`. It is loosely related to the constraints, and can work without them. It can be helpful for queries that have filtering by rarely met condition. 
E.g. we have table `hits` and column `BrowserName String` and we want to calculate aggregations for some rare browser in it  often. Then we can add and index:
```sql
ALTER TABLE hits ADD INDEX ind_hyp (BrowserName != 'RareBrowserName') TYPE hypothesis GRANULARITY 1
```
After that in queries like `SELECT count() FROM hits WHERE BrowserName = 'RareBrowserName'` granules in which condition `BrowserName = 'RareBrowserName'` isn't met in the whole granule wil be skipped.

More examples: https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/01624_soft_constraints.sh